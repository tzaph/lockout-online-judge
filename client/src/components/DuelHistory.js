import React, { useEffect, useState } from "react";
import { useAuth } from "../contexts/AuthContext";
import { getDatabase, update, ref, get, child } from "firebase/database";

export default function DuelHistory() {
  const [data, setData] = useState({});
  const [arr, setArr] = useState([]);
  const [msgArr, setMsgArr] = useState([]);
  const [loading, setLoading] = useState(false);
  const [emptyHistory, setEmptyHistory] = useState(false);
  const [msg, setMsg] = useState("");
  const currentUser = useAuth()?.currentUser;

  const Getdata = () => {
    get(child(ref(getDatabase()), "users/" + currentUser.uid))
      .then((snapshot) => {
        if (snapshot.exists()) {
          setData(snapshot.val());
        }
      })
      .catch((err) => {
        console.error(err);
      });
  };
  useEffect(() => {
    Getdata();
  }, []);

  const refreshPage = async () => {
    setLoading(true);
    setTimeout(() => setLoading(false), 2000);

    let li = data.duelHistory;
    const tmparr = [];
    const tmpmsgarr = [];

    if (li == undefined) {
      await update(ref(getDatabase(), "users/" + currentUser.uid), {
        duelHistory: tmparr,
      });
      setEmptyHistory(true);
      setMsg("You do not have any recent duels");
      return;
    }
    let len = li.length;

    if (len == 0) {
      setEmptyHistory(true);
      setMsg("You do not have any recent duels");
      return;
    }

    for (let i = 0; i < len; i++) {
      if (i == 0 || li[i].roomCode != li[i - 1].roomCode) {
        tmparr.push(li[i]);
        tmpmsgarr.push(
          li[i].duelType + " " + li[i].roomCode + " vs. " + li[i].opponent.name
        );
      }
    }
    await update(ref(getDatabase(), "users/" + currentUser.uid), {
      duelHistory: tmparr,
    });
    setArr(tmparr);
    tmpmsgarr.reverse();
    setMsgArr(tmpmsgarr);
  };

  return (
    <div className="profile-container">
      <button disabled={loading} onClick={refreshPage}>
        Refresh Duel History List
      </button>
      <p>{emptyHistory ? msg : null}</p>
      <table>
        {msgArr.map((val) => {
          return <p>{val}</p>;
        })}
      </table>
    </div>
  );
}
