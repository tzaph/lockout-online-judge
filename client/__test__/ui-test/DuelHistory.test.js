import React from "react";
import renderer from "react-test-renderer";
import DuelHistory from "../../src/components/DuelHistory";
import { BrowserRouter as Router } from "react-router-dom";

describe("<DuelHistory />", () => {
  it("Renders correctly", () => {
    const tree = renderer
      .create(
        <Router>
          <DuelHistory />
        </Router>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
