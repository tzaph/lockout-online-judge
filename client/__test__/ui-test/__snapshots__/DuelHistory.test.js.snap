// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<DuelHistory /> Renders correctly 1`] = `
<div
  className="profile-container"
>
  <button
    disabled={false}
    onClick={[Function]}
  >
    Refresh Duel History List
  </button>
  <p />
  <table />
</div>
`;
